piv/key.go

```diff
@@ -667,24 +671,82 @@ func (yk *YubiKey) PrivateKey(slot Slot, public crypto.PublicKey, auth KeyAuth)
 	}
 }
 
-type keyECDSA struct {
+// ECDSAPrivateKey is used to access a ECDSA signing key.
+type ECDSAPrivateKey struct {
 	yk   *YubiKey
 	slot Slot
 	pub  *ecdsa.PublicKey
 	auth KeyAuth
 	pp   PINPolicy
 }
 
-func (k *keyECDSA) Public() crypto.PublicKey {
+// Public returns the public key associated with this private key.
+func (k *ECDSAPrivateKey) Public() crypto.PublicKey {
 	return k.pub
 }
 
-func (k *keyECDSA) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+// Sign implements crypto.Signer.
+func (k *ECDSAPrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
 	return k.auth.do(k.yk, k.pp, func(tx *scTx) ([]byte, error) {
 		return ykSignECDSA(tx, k.slot, k.pub, digest)
 	})
 }
 
+// SharedKey performs a Diffie-Hellman key agreement with the peer
+// to produce a shared secret key.
+//
+// Peer's public key must use the same algorithm as the key in
+// this slot, or an error will be returned.
+//
+// Length of the result depends on the types and sizes of the keys
+// used for the operation. Callers should use a cryptographic key
+// derivation function to extract the amount of bytes they need.
+func (k *ECDSAPrivateKey) SharedKey(rand io.Reader, peer crypto.PublicKey, opts crypto.SignerOpts) ([]byte, error) {
```

Comment:
Unless there's a good reason to keep rand, crypto.PublicKey, and opt, I'd rather drop the unused arguments and have more type safety here.

```
func (k *ECDSAPrivateKey) SharedKey(peer *ecdsa.PublicKey) ([]byte, error) {}
```

Reply:
Yeah. I have no idea what the expectation is. These seem to be offered "in case the algorithm needs them" all around the PrivateKey/PublicKey ecosystem, which makes sense for interfaces that multiple algorithms would offer.

Worst case scenario here is that piv-go ends up with a `SharedKey` method that looks different from a hypothetical future stdlib interface. It might be awkward, but it's always possible to write an adapter that drops them, in front of this type, or do a major version bump to change the implementation.

I'll rip out the unused arguments.